src/test/java/org/folio/auth/authtokenmodule/AuthTokenTest.java

```diff
@@ -1023,4 +1043,161 @@ private String getMagicPermission(String endpoint) {
+      .onComplete(context.asyncAssertSuccess())
+      .compose(x -> ts.checkTokenNotRevoked(apiToken))
+      .onComplete(context.asyncAssertFailure());
+  }
+
+  @Test
+  public void testStoreRefreshTokenSingleUse(TestContext context) {
+    var ts = new RefreshTokenStore(vertx, tenant);
+    // Create and save some tokens.
+    var rt1 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt2 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt3 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var s1 = ts.saveToken(rt1);
+    var s2 = ts.saveToken(rt2);
+    var s3 = ts.saveToken(rt3);
+
+    Async async = context.async();
+    CompositeFuture.all(s1, s2, s3)
+      .compose(a -> ts.checkTokenNotRevoked(rt2))
+      .onComplete(context.asyncAssertSuccess()) // First check should succeed.
+      .compose(a -> ts.checkTokenNotRevoked(rt2))
+      .onComplete(context.asyncAssertFailure(b ->  { // Second should fail.
+        // At this point all tokens for the user should be revoked so any subsequent
+        // checks should fail.
+        ts.checkTokenNotRevoked(rt1).onComplete(context.asyncAssertFailure(c -> {
+          ts.checkTokenNotRevoked(rt3).onComplete(context.asyncAssertFailure(d -> {
+            async.complete();
+          }));
+        }));
+      }));
+  }
+
+  @Test
+  public void testStoreCleanupExpired(TestContext context) {
+    // Create some tokens.
+    var rt1 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt2 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt3 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt4 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+    var rt5 = new RefreshToken(tenant, "jones", userUUID, "http://localhost:" + port);
+
+    // Set a few tokens' expires at time to simulate expiration.
+    var now = Instant.now().getEpochSecond();
+    rt2.setExpiresAt(now - 10); // Would have expired 10 seconds ago.
+    rt4.setExpiresAt(now - 20); // Would have expired 20 seconds ago.
+
+    Async async = context.async();
+    var ts = new RefreshTokenStore(vertx, tenant);
+
+    // Other tests could have added tokens to storage, so remove all of those first.
+    ts.removeAll()
+      .compose(x -> {
+        var s1 = ts.saveToken(rt1);
+        var s2 = ts.saveToken(rt2);
+        var s3 = ts.saveToken(rt3);
+        var s4 = ts.saveToken(rt4);
+        var s5 = ts.saveToken(rt5);
+        return CompositeFuture.all(s1, s2, s3, s4, s5);
+      })
+      .compose(y -> ts.countTokensStored(tenant))
+      .onSuccess(y -> assertThat(y, is(5)))
+      .compose(y -> ts.cleanupExpiredTokens())
+      .compose(y -> ts.countTokensStored(tenant))
+      .onSuccess(y -> assertThat(y, is(3)))
+      .onComplete(context.asyncAssertSuccess(y -> {
+        async.complete();
```

Comment:
async.complete() is not needed inside of context.asyncAssertSuccess

Reply:
See my comment above. Sometimes I need it. Sometimes I don't. Any idea what is going on?

Reply:
context.asyncAssertSuccess(x -> {
        // some tests without callback
        async.complete();
      }

This async is not needed because there is a context.asyncAssertSuccess that already contains an async, and there is no other lambda.

An async is only needed if there is no asyncAssertXXX:

    ts.checkTokenNotRevoked(rt1)
    .onComplete(a -> async.complete());

or

    ts.checkTokenNotRevoked(rt1)
    .compose(a -> ts.checkTokenNotRevoked(rt2)
                    .onComplete(x -> async.complete()))

Reply:
A gave it a try before and I still get no timeout locally. Jenkins also has no timeout: https://github.com/folio-org/mod-authtoken/pull/116

Reply:
Thank you for your branch Julian! I was having a hard time seeing what you wanted me to see. I had to stare at it for a while before I realized what was going on. I was only commenting out `async.complete()` and testing that. But I also had `Async async = context.async()`. Having `context.async()` alone, without the completion was the timeout's cause.